File: Java Version/Complex.java

public class Complex {

	public Complex(double r, double i) {
		
		real = r;
		
		imaginary = i;
		
	}
	
	public Complex add(Complex num) {
		
		double realPart = real + num.real;
		
		double imaginaryPart = imaginary + num.imaginary;
		
		return new Complex(realPart, imaginaryPart);
		
	}
	
	public Complex sub(Complex num) {
		
		double realPart = real  - num.real;
		
		double imaginaryPart = imaginary - num.imaginary;
		
		return new Complex(realPart, imaginaryPart);
		
	}
	
	public Complex mul(Complex num) {
		
		double realPart = real * num.real + (-1) * (num.imaginary * imaginary);
		
		double imaginaryPart = num.real * imaginary + real * num.imaginary;
		
		return new Complex(realPart, imaginaryPart);
		
	}
	
	public Complex div(Complex num) {
		
		double conjugateReal = num.real;

		double conjugateImaginary = (-1) * num.imaginary;

		double realNumerator = real * conjugateReal + (-1) * imaginary * conjugateImaginary;

		double imaginaryNumerator = real * conjugateImaginary + imaginary * conjugateReal;

		double denominator = (num.real * conjugateReal) + num.real * conjugateImaginary + num.imaginary * conjugateReal + (-1) * num.imaginary * conjugateImaginary;

		return new Complex(realNumerator / denominator, imaginaryNumerator / denominator);
		
	}
	
	public Complex add(double dbl) {
		
		return new Complex(dbl + real, imaginary);
		
	}
	
	public Complex sub(double dbl) {
		
		return new Complex(real - dbl, imaginary);
		
	}
	
	public Complex mul(double dbl) {
		
		return new Complex(real * dbl, imaginary * dbl);
		
	}
	
	public Complex div(double dbl) {
		
		return new Complex(real / dbl, imaginary / dbl);
		
	}
	
	public Complex subOpposite(double dbl) {
		
		Complex b = new Complex(real, imaginary);
		
		Complex c = new Complex(dbl, 0);
		
		return c.sub(b);
		
	}
	
	public Complex divOpposite(double dbl) {
		
		Complex c = new Complex(dbl, 0);
		
		Complex d = new Complex(real, imaginary);
		
		return c.div(d);
		
	}
	
	public String toString() {
		
		return "(" + real + " + " + imaginary + "i)";
		
	}
	
	private double real;
	
	private double imaginary;
	
}

File: Python Version/complex.py

class complex:
    
    def __init__(self, real, imaginary):
        
        self.real = real
        
        self.imaginary = imaginary

    def __add__(self, other):
        
        if isinstance(other, float):
            
            return complex(self.real + other, self.imaginary)
        
        elif isinstance(other, complex):
            
            return complex(self.real + other.real, self.imaginary + other.imaginary)
        
        else:
            
            raise TypeError

    def __truediv__(self, other):
        
        if isinstance(other, float):
            
            return complex(self.real / other, self.imaginary / other)
        
        elif isinstance(other, complex):
            
            conjugateReal = other.real
            
            conjugateImaginary = (-1) * other.imaginary
            
            realNumerator = self.real * conjugateReal + (-1) * self.imaginary * conjugateImaginary
            
            imaginaryNumerator = self.real * conjugateImaginary + self.imaginary * conjugateReal
            
            denominator = (other.real * conjugateReal) + (other.real * conjugateImaginary) + (other.imaginary * conjugateReal) + ((-1) * other.imaginary * conjugateImaginary)
            
            return complex(realNumerator / denominator, imaginaryNumerator / denominator)
        
        else:
            
            raise TypeError

    def __mul__(self, other):
        
        if isinstance(other, float):
            
            return complex(self.real * other, self.imaginary * other)
        
        elif isinstance(other, complex):
            
            realPart = self.real * other.real + (-1) * self.imaginary * other.imaginary
            
            imaginaryPart = other.real * self.imaginary + self.real * other.imaginary
            
            return complex(realPart, imaginaryPart)
        
        else:
            
            raise TypeError    
    
    def __sub__(self, other):
        
        if isinstance(other, float):
            
            return complex(self.real - other.real, self.imaginary)
        
        elif isinstance(other, complex):
            
            return complex(self.real - other.real, self.imaginary - other.imaginary)
        
        else:
            
            raise TypeError

    
    def __rsub__(self,other):
        
        return complex(other, 0) - self

    def __rmul__(self, other):
        
        return complex(other, 0) * self
    
    def __radd__(self, other):
        
        return complex(other, 0) + self

    def __rtruediv__(self, other):
        
        return complex(other, 0) / self

    def __str__(self):
        
        return '(' + str(self.real) + ' + ' + str(self.imaginary) + 'i)'

    

if __name__ == '__main__':
    
    a = complex(1, 2)
    
    b = complex(3, 4)

    i = 2.0

    print('%s + %s = %s' % (a, b, a + b))
    
    print('%s - %s = %s' % (a, b, a - b))
    
    print('%s * %s = %s' % (a, b, a * b))
    
    print('%s / %s = %s\n' % (a, b, a / b))
    
    
    print('%s + %s = %s' % (a, i, a + i))
    
    print('%s - %s = %s' % (a, i, a - i))
    
    print('%s * %s = %s' % (a, i, a * i))
    
    print('%s / %s = %s\n' % (a, i, a / i))

    
    print('%s + %s = %s' % (i, a, i + a))
    
    print('%s - %s = %s' % (i, a, i - a))
    
    print('%s * %s = %s' % (i, a, i * a))
    
    print('%s / %s = %s' % (i, a, i / a))











    
